File: frontend/ui/src/app/types/notification-record.ts
import {DeploymentType} from '@distr-sh/distr-sdk';

export type NotificationRecordType = 'alert' | 'warning' | 'resolved' | 'update_available' | 'new_version';
export type NotificationSourceType = 'alert' | 'application' | 'artifact';

export interface NotificationRecordDeployment {
  customerOrganizationName?: string;
  deploymentTargetName: string;
  deploymentName: string;
  currentVersionName?: string;
}

/** Everything specific to what triggered a notification, denormalized when it was sent. */
export interface NotificationRecordDetails {
  summary?: string;
  customerOrganizationName?: string;
  deploymentTargetName?: string;
  applicationName?: string;
  applicationType?: DeploymentType;
  applicationVersionName?: string;
  artifactName?: string;
  artifactVersionName?: string;
  deployments?: NotificationRecordDeployment[];
}

export interface NotificationRecord {
  id: string;
  createdAt: string;
  userAccountId?: string;
  sourceType: NotificationSourceType;
  sourceConfigurationId?: string;
  subjectId?: string;
  type: NotificationRecordType;
  details: NotificationRecordDetails;
  message: string;
}

File: api/notification_record.go
package api

import (
	"time"

	"github.com/distr-sh/distr/internal/types"
	"github.com/google/uuid"
)

type NotificationRecord struct {
	ID                    uuid.UUID                    `json:"id"`
	CreatedAt             time.Time                    `json:"createdAt"`
	UserAccountID         *uuid.UUID                   `json:"userAccountId,omitempty"`
	SourceType            types.NotificationSourceType `json:"sourceType"`
	SourceConfigurationID *uuid.UUID                   `json:"sourceConfigurationId,omitempty"`
	SubjectID             *uuid.UUID                   `json:"subjectId,omitempty"`
	Type                  types.NotificationRecordType `json:"type"`
	Details               NotificationRecordDetails    `json:"details"`
	DeliveryError         string                       `json:"deliveryError"`
}

type NotificationRecordDetails struct {
	Summary                  string  `json:"summary,omitempty"`
	CustomerOrganizationName *string `json:"customerOrganizationName,omitempty"`

	DeploymentTargetName   *string               `json:"deploymentTargetName,omitempty"`
	ApplicationName        *string               `json:"applicationName,omitempty"`
	ApplicationType        *types.DeploymentType `json:"applicationType,omitempty"`
	ApplicationVersionName *string               `json:"applicationVersionName,omitempty"`
	ArtifactName           *string               `json:"artifactName,omitempty"`
	ArtifactVersionName    *string               `json:"artifactVersionName,omitempty"`

	Deployments []NotificationRecordDeployment `json:"deployments,omitempty"`
}

type NotificationRecordDeployment struct {
	CustomerOrganizationName *string `json:"customerOrganizationName,omitempty"`
	DeploymentTargetName     string  `json:"deploymentTargetName"`
	DeploymentName           string  `json:"deploymentName"`
	CurrentVersionName       *string `json:"currentVersionName,omitempty"`
}
